vscode_extension/wolverine/src/extension.ts

```diff
@@ -0,0 +1,141 @@
+	const newCharacterLocation = numberOfNewLines === 0 ? countCharacters(textStream) + currentCharacter : 0;
+	const newLineLocation = numberOfNewLines + currentLine;
+	return { newCharacterLocation, newLineLocation };
+};
+
+const deleteRange = async (activeEditor: TextEditor, range: Range) => {
+	await activeEditor.edit(editBuilder => {
+		editBuilder.delete(range);
+	});
+};
+
+// Yacine threw most of the complexity into here.
+// Holds a buffer in a javascript array, registers a event listener on a server-sent events function, builds the buffer
+// Takes a position, and flushes the buffer on a preconfigured cron into the provided cursor position.
+const useBufferToUpdateTextContentsWhileStreamingOpenAIResponse = async (activeEditor: TextEditor, position: Position, prompt: string) => {
+	let currentCharacter = position.character;
+	let currentLine = position.line;
+	let buffer: string[] = [];
+	let doneStreaming = false;
+	const onDataFunction = (word: any) => {
+		const newContent = word.toString().split('data: ')
+			.map((line: string) => line.trim())
+			.filter((line: string) => line.length > 0)
+			.map((line: string) => JSON.parse(line).choices[0].delta.content);
+		buffer = [...buffer, ...newContent];
+	};
+	streamCompletion(prompt, onDataFunction).then(() => doneStreaming = true);
+	while (!doneStreaming || buffer.length >= 0) {
+		const word: string | undefined = buffer.shift();
+		if (word) {
+			let { newCharacterLocation, newLineLocation } = getNewCursorLocation(word, currentLine, currentCharacter);
+			const position = new Position(currentLine, currentCharacter);
+			await activeEditor.edit((editBuilder) => {
+				editBuilder.insert(position, word);
+			});
+			currentCharacter = newCharacterLocation;
+			currentLine = newLineLocation;
+		}
+		// TODO I should make this buffer flush configurable.
+		await sleep(30); 
+	}
+};
+
+const constructPrompt = async (text: string): Promise<string> => {
+	const defaultPrompt = `
+INSTRUCTIONS: 
+The following text provided has come straight from my text editor. I have an extension that can send my text output to you, often.
+I've added my instructions in the comments, please attend to the comments, and output the code.
+When you output the code, please don't add any human english language. Your entire output must not cause any errors!
+Output code that effectively does the same thing of what I sent you, my extension replaces the text that I've highlighted and sent.
+Remember, you'll have to send me any messages in comments!
+CODE:
+`;
+	const configuredPrompt = await workspace.getConfiguration().get('wolverine.prompt');
+	return (configuredPrompt || defaultPrompt) + text;
+};
+
+export async function activate(context: ExtensionContext) {
+
+	let disposable = commands.registerCommand('wolverine.directedHeal', async () => {
+		// Forgive me father..
+		openaikey = await workspace.getConfiguration().get('wolverine.UNSAFE.OpenaiApiKeySetting') || '';
```

Comment:
there is literally no good solution for this so i will have to roll my own

Reply:
i can't believe they don't provide oauth

Reply:
Hi, in this extension for example [codeGPT](https://github.com/davila7/code-gpt-docs), you can give the key with a user prompt. I don't know how it is stored but it is not hardcoded in a configfile :)

Reply:
Thank you! Do you know if the extension processes are sandboxed away from each other?

Honestly, having to copy the buffer in and out probably is same amount of security even if it _was_ sandboxed; given most users will just paste it into a text file for accessibility :P

The real answer here is either keychain (a la copilot) or oauth
The fact that the keys don't expire is the first problem